### PR TITLE
refactor: extract shared CLI helpers and add DisplayMode type

### LINKED ISSUE #136: Code review findings: Low-priority CLI improvements
## Overview

During the code review of PR #134, several low-priority issues were identified that are out of scope for that focused bug fix but worth tracking for future cleanup.

### 1. Imports Inside Functions (Pre-existing)

**Files:** `src/pivot/cli/remote.py`, `src/pivot/cli/run.py`

Several imports are placed inside functions rather than at module level, which violates the CLAUDE.md import style guideline.

**Examples in remote.py:**
```python
def remote_list() -> None:
    from pivot.remote import config as remote_config  # Should be module-level

def push(...):
    from pivot.remote import sync as transfer  # Should be module-level
    from pivot.storage import state
```

**Examples in run.py:**
```python
def _get_all_explanations(...):
    from pivot import dag, explain, parameters, project  # Line 66

def _run_with_tui(...):
    import multiprocessing as mp  # Line 101
    from pivot import dag, project
    from pivot.tui import run as run_tui
```

**Recommendation:** Move all imports to module level for consistency.

---

### 2. Other CLI Files Not Using `pivot_command` Decorator

**Files:** `status.py`, `history.py`, `metrics.py`, `plots.py`

These files still use older patterns instead of the new `@pivot_command()` decorator:

| File | Current Pattern |
|------|-----------------|
| `status.py` | `@click.command()` + manual `_ensure_stages_registered()` |
| `history.py` | `@click.command()` directly |
| `metrics.py` | `@with_error_handling` + manual `ensure_stages_registered()` |
| `plots.py` | `@with_error_handling` + manual `ensure_stages_registered()` |

**Note:** `metrics.py` and `plots.py` are group subcommands, so they can't use `@pivot_command()` directly (per CLAUDE.md). However, `status.py` and `history.py` could be migrated.

**Recommendation:** Migrate `status.py` and `history.py` to use `@pivot_command()`.

---

### 3. Redundant Comments in run.py (Pre-existing)

Two comments violate the "comments explain WHY not WHAT" guideline:

**Line 321:**
```python
pass  # Normal exit via Ctrl+C
```
The comment states the obvious—the `except KeyboardInterrupt:` block makes this clear.

**Line 334:**
```python
# Disable TUI when JSON output is requested
```
The code `use_tui = ... and not as_json` already expresses this clearly.

**Recommendation:** Remove both comments.

---

### 4. Duplicate `ensure_stages_registered()` Functions

Both `run.py` and `status.py` define their own version:
- `run.py`: `ensure_stages_registered()` (public)
- `status.py`: `_ensure_stages_registered()` (private)

Both do exactly the same thing.

**Recommendation:** Keep only `run.py`'s version and import it where needed (already done for group subcommands per CLAUDE.md).

---

## Priority

All items are **low priority** and can be addressed opportunistically during related work.

## Related

- PR #134 - Auto-discovery fix that prompted this review
- Issue #135 - Bug discovery during integration testing

### PR DESCRIPTION
**Pull request by sjawhar**
## Summary

This PR refactors the CLI code to extract shared utilities and improve type safety:

- **New shared helpers module** (`src/pivot/cli/helpers.py`):
  - `validate_stages_exist()` - Consistent stage validation using `StageNotFoundError`
  - `make_progress_callback()` - Factory for transfer progress display
  - `print_transfer_errors()` - Consistent error display with truncation

- **DisplayModeType custom Click ParamType**:
  - Converts string to `DisplayMode` enum at parse time
  - Provides shell completion for tui/plain values
  - Uses proper `@override` decorators

- **Code consolidation**:
  - `run.py` uses `DisplayModeType` instead of `click.Choice`
  - `status.py` uses shared helper (removes duplicate `_validate_stages`)
  - `remote.py` uses shared progress and error helpers

## Test plan

- [x] All existing tests pass (2009 passed)
- [x] `ruff check` passes
- [x] `basedpyright` passes with no errors

Closes #136

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by sjawhar** [comment]
### Code review

No issues found. Checked for bugs and CLAUDE.md compliance.

Generated with [Claude Code](https://claude.ai/code)

**Review by sjawhar — COMMENTED** [review]
## Review Summary

**Recommendation: Request Changes** - This PR makes good progress on the code cleanup requested in issue #136, but has one blocking issue (benchmark file in PR) and a few important gaps.

## What Works Well

- **Module-level imports:** The PR correctly moves all imports to module level in `run.py` and `remote.py`, addressing one of the key issues from #136
- **`pivot_command` migration:** Successfully migrates `status.py` and `history.py` to use `@pivot_command()` decorator, as requested
- **Shared helpers extraction:** `cli/helpers.py` consolidates duplicate code for stage validation, progress callbacks, and error printing
- **DisplayModeType:** Clean implementation of custom Click ParamType with proper `@override` decorators and shell completion
- **Code quality:** All files pass `ruff check` and `basedpyright` with no errors or warnings
- **Tests pass:** All 66 related tests pass

## Blocking Issues

1. **[BLOCKING] `benchmark_no_commit.py` included in PR** - This 300-line benchmark script is included in the diff. The filename itself suggests it should not be committed ("no_commit"). This file:
   - Has a module-level docstring (violates CLAUDE.md: "No module-level docstrings")
   - Uses `sys.path.insert()` for imports (non-standard)
   - Is not related to the refactoring scope of issue #136
   - Should be added to `.gitignore` or removed from the PR

## Important Issues

1. **[IMPORTANT] Missing tests for `cli/helpers.py`** - The new helpers module has no unit tests. Per CLAUDE.md, the project requires 90%+ coverage. The three functions (`validate_stages_exist`, `make_progress_callback`, `print_transfer_errors`) should have direct unit tests. While they are tested indirectly through CLI tests, explicit tests would verify edge cases like:
   - `validate_stages_exist` with None input
   - `validate_stages_exist` with empty list
   - `make_progress_callback` output format
   - `print_transfer_errors` truncation logic

2. **[IMPORTANT] Docstring style in helpers.py** - Lines 27-35 and 44-50 use Args/Returns sections in docstrings, which violates CLAUDE.md ("skip Args/Returns if type hints make it obvious"). The type hints already make the parameters clear.

3. **[IMPORTANT] Issue #136 items not fully addressed:**
   - Issue mentions removing redundant comments in run.py (line 321: `pass  # Normal exit via Ctrl+C`). This comment was removed - good!
   - But the issue also mentions line 334's comment "Disable TUI when JSON output is requested" - this was also removed, which is correct.

## Additional Suggestions

1. **[SUGGESTION] `DisplayModeType` could live in a separate types module** - Currently it's defined in `run.py`. If other CLI commands need display mode options, consider moving it to `cli/types.py` or similar. However, this is fine for now since only `run` uses it.

2. **[NITPICK] Line 63 in run.py:** `name: str = "display_mode"` - The `: str` annotation is redundant since the string literal already implies the type. Could be `name = "display_mode"`.

## Testing Notes

- All 66 CLI-related tests pass
- New helpers module indirectly tested through CLI tests for push/pull/status
- Direct unit tests for `helpers.py` would improve confidence and coverage

## Next Steps

1. **Remove `benchmark_no_commit.py` from the PR** (blocking)
2. **Add tests for `cli/helpers.py`** (important for coverage)
3. **Simplify docstrings in helpers.py** to match project style (important)

File: benchmark_no_commit.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
+#!/usr/bin/env python3
```
Comment (sjawhar) [inline benchmark_no_commit.py]: **[BLOCKING]** This benchmark file should not be included in the PR. The filename 'no_commit' suggests this is intentionally meant to be excluded. Additionally:

- Has a module-level docstring (violates CLAUDE.md)
- Uses `sys.path.insert()` for imports
- Not related to issue #136 scope

Please remove this file from the PR or add it to `.gitignore`.

File: src/pivot/cli/helpers.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+def make_progress_callback(action: str) -> Callable[[int], None]:
+    """Create a progress callback for file transfer operations.
+
+    Args:
+        action: The action being performed (e.g., "Uploaded", "Downloaded")
+
+    Returns:
+        A callback function that displays progress using carriage return
+    """
```
Comment (sjawhar) [inline src/pivot/cli/helpers.py]: **[IMPORTANT]** Per CLAUDE.md docstring guidelines: 'skip Args/Returns if type hints make it obvious'. The type hints on line 27 (`action: str`) and the return type `Callable[[int], None]` already make this clear. Suggest simplifying to:

```python
def make_progress_callback(action: str) -> Callable[[int], None]:
    """Create a progress callback for file transfer operations."""
```

File: src/pivot/cli/run.py [hunk 3]
```diff
@@ -43,17 +57,41 @@ def ensure_stages_registered() -> None:
             raise click.ClickException(str(e)) from e
 
 
+class DisplayModeType(click.ParamType):
+    """Click parameter type that converts string to DisplayMode enum."""
+
+    name: str = "display_mode"
```
Comment (sjawhar) [inline src/pivot/cli/run.py]: **[NITPICK]** The `: str` annotation is redundant since the string literal `"display_mode"` already implies the type. Could simplify to:

```python
name = "display_mode"
```

However, this is a minor style preference and not blocking.

File: src/pivot/cli/helpers.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    Args:
+        errors: List of error messages
+        max_shown: Maximum number of errors to display before truncating
+    """
+    if not errors:
+        return
+    for err in errors[:max_shown]:
+        click.echo(f"  Error: {err}", err=True)
+    if len(errors) > max_shown:
+        click.echo(f"  ... and {len(errors) - max_shown} more errors", err=True)
```
Comment (sjawhar) [inline src/pivot/cli/helpers.py]: **[IMPORTANT]** This new module needs unit tests to meet the project's 90%+ coverage requirement. Please add tests in `tests/cli/test_cli_helpers.py` covering:

1. `validate_stages_exist`:
   - With `None` input (should return silently)
   - With empty list (should return silently)
   - With valid stage names (should not raise)
   - With unknown stage names (should raise `StageNotFoundError`)

2. `make_progress_callback`:
   - Returns a callable
   - Callable produces expected output format

3. `print_transfer_errors`:
   - Empty list (no output)
   - List within max_shown limit
   - List exceeding max_shown (truncation message)
